### DISCUSSION TITLE
Adding a table

### DISCUSSION THREAD

**Discussion by Dreams005** [TomSchimansky/CustomTkinter#431] [Q&A] [answered]
Hey, pretty self-explanatory. I've been trying to find out how to add a table using customtkinter but I can't find anything related to it.
Could anyone show me how to implement one please?

Edit: This is my first time ever using anything related to python GUIs

**Comment by Dreams005** [comment, marked as answer]
nvm i just worked around this limitation by adding an old tkinter table and changing it to look modern
![image](https://user-images.githubusercontent.com/100276137/188245770-a03e6608-cd36-4732-bab7-518f6408437b.png)

**Reply by IndrajeethY** [reply]
how u did this?

**Reply by Dreams005** [reply]
used a normal tkinter treeview and changed the style

**Reply by IndrajeethY** [reply]
can u share ur style code formats pls

**Reply by Dreams005** [reply]
sure one second

**Reply by Dreams005** [reply]
`    style = ttk.Style()
        
                style.theme_use("default")
        
                style.configure("Treeview",
                                background="#2a2d2e",
                                foreground="white",
                                rowheight=25,
                                fieldbackground="#343638",
                                bordercolor="#343638",
                                borderwidth=0)
                style.map('Treeview', background=[('selected', '#22559b')])
        
                style.configure("Treeview.Heading",
                                background="#565b5e",
                                foreground="white",
                                relief="flat")
                style.map("Treeview.Heading",
                          background=[('active', '#3484F0')])`

**Reply by FidelMhaule** [reply]
> ```
> `    style = ttk.Style()
>     
>             style.theme_use("default")
>     
>             style.configure("Treeview",
>                             background="#2a2d2e",
>                             foreground="white",
>                             rowheight=25,
>                             fieldbackground="#343638",
>                             bordercolor="#343638",
>                             borderwidth=0)
>             style.map('Treeview', background=[('selected', '#22559b')])
>     
>             style.configure("Treeview.Heading",
>                             background="#565b5e",
>                             foreground="white",
>                             relief="flat")
>             style.map("Treeview.Heading",
>                       background=[('active', '#3484F0')])`
> ```

Hi

I want to add the treeview. How do I use the Tkinter module in Custom Tkinter?

**Reply by ryukaizen** [reply]
Based

**Reply by pinkurauchin** [reply]
Hi! I'm using this method to show data on my application, but i'm having this problem:
<img width="671" alt="image" src="https://user-images.githubusercontent.com/93226265/218101129-bad61427-e5c9-4834-8857-f401578133ea.png">

I dont know why, but so many time it appears with that blank spaces before and after the table, do u know why it should be happening? 
Thanks

**Reply by Dreams005** [reply]
@pinkurauchin , I have not coded in forever so I have absolutely no idea. But, it might be that the size of that frame is too large, try decreasing the size of it.

**Reply by febin-j** [reply]
Hi @Dreams005 how did you get your table to fit inside of your frame? When I try to do that my frame just dissapears

**Comment by RobbieKhan** [comment]
Best table widget for tkinter that I have ever seen is tksheet packet.
You can interact with all table cells, select\deselect them, copy, delete, move columns and so on. Minimal example is below.

```
import tkinter.ttk as ttk
import tksheet
...

class UI_TabTable(ttk.Frame):
    def __init__(self, main_window, **kw):
            self.table = tksheet.Sheet(self, font=GUI_FONT_CONTENT, header_font=GUI_FONT_CONTENT,
                                       popup_menu_font=GUI_FONT_CONTENT, table_bg=GUI_BG_COLOR)
            self.table.headers(['Timestamp', 'Temperature, °C', 'Humidity, %', 'Annotation'])
            self.table.column_width(column=0, width=180, only_set_if_too_small=False)
            self.table.column_width(column=2, width=100, only_set_if_too_small=False)
            self.table.column_width(column=3, width=180, only_set_if_too_small=False)
            self.table.pack(expand=True, fill='both')
            self.table.enable_bindings('all')
            self.table.disable_bindings(['rc_insert_column', 'rc_delete_column', 'edit_cell', 'delete', 'paste',
                                         'cut', 'rc_delete_row', 'rc_insert_row'])

    def add(self, timestamp, temperature, humidity, is_tampered, is_bookmark):
        self.table.insert_row(values=(timestamp, temperature, humidity, is_tampered, is_bookmark),
                              idx='end', add_columns=False)

    def clear(self):
        self.table.select_all(redraw=False, run_binding_func=True)
        rows_number = len(self.table.get_selected_rows(get_cells_as_rows=True))
        self.table.MT.del_row_positions(idx=0, numrows=rows_number, deselect_all=True)
```

How it looks:
![image](https://user-images.githubusercontent.com/53120051/210140397-77a6d197-4d41-447d-a47f-5c034056afa9.png)

**Reply by Nexa10** [reply]
@Dreams005 Please can you share the code for the table

**Reply by sathyarajshetigar** [reply]
+1

**Reply by Dreams005** [reply]
Sorry @Nexa10 I just saw this now thanks to @sathyarajshetigar. I haven't coded in forever so I'm not sure if this is up to date anymore but this is what I used

`       
        self.grid_columnconfigure(1, weight=1)
        self.grid_rowconfigure(0, weight=1)

        self.frame_left = customtkinter.CTkFrame(master=self,
                                                 width=200,
                                                 corner_radius=0)
        self.frame_left.grid(row=0, column=0, sticky="nswe")

        self.frame_right = customtkinter.CTkFrame(master=self)
        self.frame_right.grid(row=0, column=1, sticky="nswe", padx=15, pady=15)

        self.frame_left.grid_rowconfigure(0, minsize=10)
        self.frame_left.grid_rowconfigure(8, minsize=20)
        self.frame_left.grid_rowconfigure(11, minsize=10)

        self.add_menu_display211 = customtkinter.CTkFrame(master=self.frame_right,
                                                          corner_radius=15,
                                                          height=400,
                                                          width=600)
        self.add_menu_display211.grid(pady=15, padx=15, sticky="nws")

        columns = ('id', 'item', 'date_added', 'date_of_manufacture')

        self.table = ttk.Treeview(master=self.add_menu_display211,
                                  columns=columns,
                                  height=17,
                                  selectmode='browse',
                                  show='headings')

        self.table.column("#1", anchor="c", minwidth=50, width=50)
        self.table.column("#2", anchor="w", minwidth=220, width=220)
        self.table.column("#3", anchor="c", minwidth=120, width=120)
        self.table.column("#4", anchor="c", minwidth=120, width=120)

        self.table.heading('id', text='ID')
        self.table.heading('item', text='Item')
        self.table.heading('date_added', text='Date Added')
        self.table.heading('date_of_manufacture', text='Date of Manufacture')

        self.table.grid(row=0, column=0, sticky='nsew', padx=10, pady=10)

        self.table.bind('<Motion>', 'break')`

**Reply by Nexa10** [reply]
thanks

**Comment by sathyarajshetigar** [comment]
I implemented the treeview and its working fine in 1080P monitor but if I open it in a 4K monitor scaling gets messed up
this is how it looks in 4k
![image](https://user-images.githubusercontent.com/42057975/216049901-b45f783a-c7da-4be0-9f2d-c81eff5c06d0.png)
and this is how it looks in 1080p screen
![image](https://user-images.githubusercontent.com/42057975/216050055-ae886762-7a27-4ee8-9b67-e9c458ae8262.png)
any idea how to fix the scaling

**Comment by Akascape** [comment]
@Dreams005 Here is a table widget for customtkinter: https://github.com/Akascape/CTkTable
![Screenshot](https://user-images.githubusercontent.com/89206401/233420929-bf210cb3-5b5f-49b2-ba7a-f01d187e72cf.jpg)

**Reply by sathyarajshetigar** [reply]
Thanks. Let me try it out

**Reply by santhoshu2903** [reply]
how to get the focus from this tables, like get object of hovered row

**Reply by Akascape** [reply]
@santhoshu2903 Check this out: https://github.com/Akascape/CTkTable/issues/6 and https://github.com/Akascape/CTkTable/issues/29#issuecomment-1627278004

**Comment by khawajaRafey** [comment]
The [CTkTable](https://github.com/Akascape/CTkTable) is good and has many good functions. But if you have more records, like I tested it by loading a thousand records from the database into the table, the app took 5-8 seconds to start and then a laggy, stucked black screen. So if you have a few records its good. Otherwise in case of more records it's not recommended. Thanks!

**Reply by nadershehata95** [reply]
What's recommended for more many records then please?

**Reply by dipeshSam** [reply]
@nadershehata95 You can use **[tksheet](https://github.com/ragardner/tksheet)** . It draws the table on Tkinter Canvas instead of creating the table using widgets resulting in smoother performance and you can handle tons of data efficiently. It is also interactive and has so many advanced features. You can read the [documentation](https://github.com/ragardner/tksheet/wiki/Version-7#table-of-contents).

The other similar library is **[Pandastable or Tkinter Table](https://github.com/dmnfarrell/pandastable)**. It is even more versatile, check it out.

Regards.
